File: app02_angular/backend/src/main/java/com/threatview/entity/Threat.java
package com.threatview.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.List;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "threat")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Threat {

    @Id
    @GeneratedValue
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "framework_id", nullable = false)
    private Framework framework;

    @Column(nullable = false, length = 64)
    private String code;

    @Column(nullable = false)
    private String title;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 16)
    private Severity severity;

    @Column(length = 128)
    private String category;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "attack_vector", columnDefinition = "TEXT")
    private String attackVector;

    @Column(name = "attack_surface", columnDefinition = "TEXT")
    private String attackSurface;

    @Convert(converter = StrideSetConverter.class)
    @Column(length = 32)
    private Set<StrideCategory> stride;

    @Convert(converter = StringListConverter.class)
    @Column(name = "cve_references", columnDefinition = "TEXT")
    private List<String> cveReferences;

    @Convert(converter = StringListConverter.class)
    @Column(columnDefinition = "TEXT")
    private List<String> tags;
}

File: app10_csharp_react/backend/src/SharpGuard.Api/Entities/Threat.cs
namespace SharpGuard.Api.Entities;

public class Threat
{
    public Guid Id { get; set; }
    public Guid FrameworkId { get; set; }
    public Framework Framework { get; set; } = null!;

    public required string Code { get; set; }
    public required string Title { get; set; }
    public Severity Severity { get; set; }
    public string? Category { get; set; }
    public string? Description { get; set; }
    public string? AttackVector { get; set; }
    public string? AttackSurface { get; set; }

    // Comma-joined TEXT columns in Postgres - same deliberate Phase-1
    // shortcut as app01's StringListConverter/StrideSetConverter, not a
    // native array/JSON column type. Kept as raw strings at the entity/query
    // layer (not List<string> via a value converter) on purpose: EF Core
    // can't translate LINQ predicates against a converted collection back to
    // SQL on the underlying scalar column, so filtering (stride/tag query
    // params) needs the raw string here - conversion to List<string> for the
    // response body happens in the DTO mapping instead (Dtos/ThreatDtos.cs).
    public string? Stride { get; set; }
    public string? CveReferences { get; set; }
    public string? Tags { get; set; }
}
